apps/rush-lib/src/logic/InstallManager.ts

```diff
@@ -942,14 +942,17 @@ export class InstallManager {
             undefined,
             false, () => {
               if (this._rushConfiguration.packageManager === 'pnpm') {
-                // If there is a failure in pnpm, it is possible that it left the
-                // store in a bad state. Therefore, we should clean out the store
-                // before attempting the install again.
-
                 console.log(colors.yellow(`Deleting the "node_modules" folder`));
                 this._commonTempFolderRecycler.moveFolder(commonNodeModulesFolder);
-                console.log(colors.yellow(`Deleting the "pnpm-store" folder`));
-                this._commonTempFolderRecycler.moveFolder(this._rushConfiguration.pnpmStoreFolder);
+
+                // pnpm-store is managed by PNPM internally, so leave it as is for the retry.
+                // This also ensures that packages that have already been downloaded need not be
+                // downloaded again, thereby potentially increasing the chances of a subsequent successful
+                // install. If the installation fails again, rush purge may be used to clear
+                // the common/temp folder.
+
+                console.log(colors.yellow(`Note: The pnpm-store was not deleted. ` +
```

Comment:
It's conceivable that a pnpm crash could leave its store in a bad state and I assume that pnpm would be able to recover from it in its next run. But to be more careful, maybe we could use the file marker logic here. Create the marker file before the installation starts, and delete it when the installation succeeds. Next time rush install is invoked, rush would know that the cache (node_modules / pnpm-store) needs to be cleared first if it finds the marker.

This approach would mean that a retry wouldn't clear the cache, but if that also fails, the subsequent "rush install" would find the marker and would clear the cache before proceeding with the install.

Reply:
With Azure DevOps, if someone pushes an update to their PR branch while `rush install` is running, then the process will be immediately terminated, in order to restart the build with the latest commit.  We would not want to clear the cache in this case, since everything is (presumably) still in good shape.

Whereas when `pnpm install` reports an error, that's the situation where we start to wonder whether its cache folder is corrupted.  The right solution is to examine the error codes/messages from PNPM.

But if that's too expensive, then we need to rely on heuristics.  Until we see evidence to the contrary, I'm willing to believe that PNPM will never corrupt its cache folder /provided/ that it was created by the same major/minor version of PNPM.  I'm much less confident about compatibility with future/past versions, since it's a huge test matrix, they have limited developer resources.

Reply:
To make this feedback more concrete:

1. For a quick short term fix, if an error occurs, purge the folder ONLY if the major/minor version number for PNPM is different from the version of the last successful install.  This information could be stored in a marker file of some kind.

2. Longer term, we need to figure out how to obtain error codes from PNPM when it fails.  Perhaps this means invoking it via the API instead of the CLI.

Reply:
An easier way to achieve the above goals would be to delete the store if and only if all the installation attempts fail. In this approach:
1) The store is retained during retries. This helps in the case of a retry if the previous installation attempt had failed for n/w reasons. This also helps in the case of a canceled PR build.
2) The store is deleted if all the installation attempts fail. Now, a following install can start things afresh. This will also help in the case of the store running into an irrecoverable state due to running a version of PNPM, different than what created the store originally..

This approach ensures that rush won't run into a situation in which the store needs to be manually deleted. 
Something like:

```typescript
try {
  executeWithRetries(install, MAX_INSTALL_ATTEMPTS, () => {
    // attempt failure callback
    // Don't delete the pnpm store
    
    delete(commonNodeModulesFolder);
  });
} catch (error) {
  // All the installation attempts fail
  if (packageManager === PNPM) {
    delete(store);
  }
  
  throw error;
}
```